### PR TITLE
build: Run functional tests and benchmarks under the undefined behaviour sanitizer (UBSan)

### PR DESCRIPTION
**Pull request by practicalswift**
Run functional tests and benchmarks under the undefined behaviour sanitizer (UBSan).

This will make Travis automatically detect issues such as:
* #14242: Avoid triggering undefined behaviour (`std::memset(nullptr, 0, 0)`) if an invalid string is passed to `DecodeSecret(...)`
* #14239: Avoid dividing by zero (undefined behaviour) in `EstimateMedianVal` (policy)/`ConnectTip` (validation)/`CreateTransaction` (wallet)
* #13546: wallet: Avoid potential use of uninitialized value `bnb_used` in `CWallet::CreateTransaction(...)`

Addresses issue #14059.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -99,24 +99,15 @@ jobs:
...
         BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-glibc-back-compat --enable-reduce-exports --enable-debug CXXFLAGS=\"-g0 -O2\""
-# x86_64 Linux (no depends, only system libs)
-    - stage: test
-      env: >-
-        HOST=x86_64-unknown-linux-gnu
-        PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
-        NO_DEPENDS=1
-        GOAL="install"
-        BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER"
-# x86_64 Linux (sanitizers)
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply .travis.yml]: It is not removed - the two jobs are merged in accordance to this comment:

```
Disabled for now, can be combined with the other x86_64 linux NO_DEPENDS job when functional tests pass the sanitizers
```

Let me know if you see any technical reasons for not doing that job merge :-)

**Comment by practicalswift** [comment]
@MarcoFalke Would you mind reviewing? :-)

**Comment by practicalswift** [comment]
Rebased!

File: .travis.yml [hunk 2]
```diff
@@ -111,15 +111,15 @@ jobs:
...
         BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER"
-# x86_64 Linux (sanitizers)
+# x86_64 Linux (no depends, only system libs, sanitizers: undefined (UBSAN) + integer)
     - stage: test
       env: >-
         HOST=x86_64-unknown-linux-gnu
         PACKAGES="clang python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
         NO_DEPENDS=1
-        RUN_FUNCTIONAL_TESTS=false # Disabled for now, can be combined with the other x86_64 linux NO_DEPENDS job when functional tests pass the sanitizers
+        RUN_BENCH=true
```
Comment (MarcoFalke) [inline .travis.yml]: why?

Also could be combined with the above job?

Comment (practicalswift) [reply .travis.yml]: Incorrect merge resolved. Please re-review :-)

File: .travis.yml [hunk 3]
```diff
@@ -103,23 +103,15 @@ jobs:
...
-        BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER"
-# x86_64 Linux (sanitizers)
+# x86_64 Linux (no depends, only system libs, sanitizers: undefined (UBSAN) + integer)
     - stage: test
       env: >-
         HOST=x86_64-unknown-linux-gnu
         PACKAGES="clang python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev"
         NO_DEPENDS=1
-        RUN_FUNCTIONAL_TESTS=false # Disabled for now, can be combined with the other x86_64 linux NO_DEPENDS job when functional tests pass the sanitizers
+        RUN_FUNCTIONAL_TESTS=true
```
Comment (MarcoFalke) [inline .travis.yml]: The default is already true. Could remove this line?

Comment (practicalswift) [reply .travis.yml]: Done! Please re-review :-)

**Pull request merged**